### ISSUE TITLE
[🐛 Bug]: Not-found page doesn't work with last vercel version v31.2.0 in edge runtime

### ISSUE THREAD

**Issue by AlexandrNetsvetov** [cloudflare/next-on-pages#413]
### next-on-pages environment related information

System:
        Platform: darwin
        Arch: arm64
        Version: Darwin Kernel Version 22.4.0: Mon Mar  6 21:00:41 PST 2023; root:xnu-8796.101.5~3/RELEASE_ARM64_T8103
        CPU: (8) arm64 Apple M1
        Memory: 16 GB
        Shell: /bin/zsh
Binaries:
        Node: 18.15.0
        Yarn: 1.22.17
        npm: 9.5.0
        pnpm: 8.6.5
Package Manager Used: yarn (classic)
Relevant Packages:
        @cloudflare/next-on-pages: 1.4.0
        vercel: N/A
        next: N/A

### Description

Hi guys.
I tried to run `npx @cloudflare/next-on-pages` on Friday and got this error. (Everything was fine a day before).

<img width="1057" alt="Screenshot 2023-07-30 at 19 50 32" src="https://github.com/cloudflare/next-on-pages/assets/137889240/b6f17031-6488-44d4-a5fd-bebdc04cf85f">

Vercel version had been up to 31.2.0 this day so maybe that was the problem.

When I tried to add `export const runtime = "edge"; ` to `not-found.tsx` page I've got this error:

<img width="927" alt="Screenshot 2023-07-30 at 19 57 47" src="https://github.com/cloudflare/next-on-pages/assets/137889240/384f01c3-c72f-468c-a2d6-d989e26ac538">

not-found.tsx

```javascript
export const runtime = "edge";

export default async function NotFoundPage() {
  return (
    <html lang="en">
      <body>
        <div>test</div>
      </body>
    </html>
  );
}

### Reproduction

_No response_

### Pages Deployment Method

None

### Pages Deployment ID

_No response_

### Additional Information

_No response_

### Would you like to help?

- [ ] Would you like to help fixing this bug?

**Comment by BlackBerry009** [comment]
Yes. I have the same problem.

**Comment by dario-piotrowicz** [comment]
Hi @AlexandrNetsvetov thank you very much for the issue 🙂 

The second part of the issue is a duplicate of #288

The first could be a very similar issue, it's interesting that it worked on an older Vercel CLI but not on a new one, I'll investigate it

if you could share your code that'd be very helpful 🙏 , if that's not possible, anyways I'd just like to double-check, am I correct in thinking that you're using app directory? (as in pages you'd have `404.tsx/jsx` instead I think?)

___

PS: @BlackBerry009, you as well, if you can share your code that could be very helpful 🙂 🙏  (if you experienced the first part of the issue this is the right place, if you experienced the second part please share info in #288)

**Comment by AlexandrNetsvetov** [comment]
@dario-piotrowicz Thanks for your answer.

Yes, I'm using app directory.

I can't share my code but I'm ready to provide you all details that you need.

PS: I've added `export const runtime = "edge";` field only because my build had been fallen with first error message. My not-found page used to worked without edge runtime for vercel 31.1.1.

**Comment by dario-piotrowicz** [comment]
Thanks a lot for the info @AlexandrNetsvetov 😄 

I'll investigate and if more info is needed I'll let you know 🙂 🙏

**Comment by dario-piotrowicz** [comment]
Minimal Reproduction: https://github.com/dario-piotrowicz/next-on-pages-app-not-found-vercel-31-issue-reproduction

**Comment by emrecoban** [comment]
My app throw the same error.

Next v13.4.13
cloudflare/next-on-pages v1.5.0

**Comment by dario-piotrowicz** [comment]
I've just merged #418 which introduces a temporary workaround for this issue, it allows the use of a not-found page as long as it doesn't contain runtime logic

It is not ideal but the best we can do for now, I've marked this issue as `blocked by external` as we'll have to investigate if maybe Vercel/Next can the generation of serverless functions for such routes on their end

**Comment by emrecoban** [comment]
I've tried that solution, but the issue still occurs @dario-piotrowicz.

```json
{
  "devDependencies": {
    "@cloudflare/next-on-pages": "https://prerelease-registry.devprod.cloudflare.dev/next-on-pages/runs/5786546906/npm-package-next-on-pages-418",
  }
}
```

Build logs in no edge runtime:
```bash
22:21:28.531 | ⚡️The following routes were not configured to run with the Edge Runtime:
22:21:28.531 | ⚡️ - /_not-found
22:21:28.531 | ⚡️
22:21:28.532 | ⚡️Please make sure that all your non-static routes export the following edge runtime route segment config:
22:21:28.532 | ⚡️ export const runtime = 'edge';
```

Build logs in edge runtime:
```bash
22:24:11.400 | ⚡️The following routes were not configured to run with the Edge Runtime:
22:24:11.401 | ⚡️  - /_error
22:24:11.401 | ⚡️
22:24:11.401 | ⚡️Please make sure that all your non-static routes export the following edge runtime route segment config:
22:24:11.401 | ⚡️ export const runtime = 'edge';
```

**Comment by dario-piotrowicz** [comment]
@emrecoban are you sure the prerelease got installed correctly? 😓 

can you double check and make sure that next-on-pages tells you that it's using the prerelease version?
(like [this](https://github.com/dario-piotrowicz/next-apps-for-testing/actions/runs/5813994024/job/15762765086#step:8:28))

I just used the prerelease to deploy [this application](https://github.com/dario-piotrowicz/next-apps-for-testing/tree/master/apps/app-dir-not-founds) which is getting [built correctly](https://github.com/dario-piotrowicz/next-apps-for-testing/actions/runs/5813994024/job/15762765086#step:8:305) and working as expected:
 https://ed46310b.next-apps-for-testing.pages.dev/

(If you confirm that you're indeed using the prerelease, can you spot something different between your application and mine that could make yours not work?)

**Comment by emrecoban** [comment]
Finally I'm here. Okay, I updated next-on-pages, and then it worked. But I had to add Vercel@30.0.0 package, or something (I'm not sure). Because I couldn't update the package for a while as follow (The message I gave up to send).
_

> It seems that the reason I got the error is cloudflare/next-on-pages not updated.
> 
> I'm investigating how to update the package to pre-version, but  pnpm throws that error: `No matching version found for @cloudflare/next-on-pages@0.0.0-d1839d6`. I didn't get why pnpm throws this one.
> 
> I've checked the versions of cloudflare/next-on-pages, and have not found  the commit `d1839d6`.

_

**The next issue:** I get that error sometimes while I refresh the `not-found.js`:
![instead-not-found](https://github.com/cloudflare/next-on-pages/assets/4332481/7ba90f0e-65b2-4318-8f05-a4a9512eceac)
![404](https://github.com/cloudflare/next-on-pages/assets/4332481/59ced557-d105-448a-afe6-5244a1d46d35)


Sometimes not-found, sometimes that error.
My guess is that Next.js doesn't catch not-found.js (sometimes, IDK why sometimes) and try to run the server component (fetching data).

**Comment by dario-piotrowicz** [comment]
@emrecoban awesome! no problem! thanks so much for confirming that's all good now (or at least as good as it can be given our limitations 😅) ❤️

**Comment by vans163** [comment]
same error, so annoying. remix just worked out of the box. I will side with joe armstrong the creator of erlang here RIP, when he said NPM and the node ecosystem is simply wack, taking something that should be smoothly and easily and adding a layer of 3 PhDs required just to deploy a project.

**Comment by vans163** [comment]
```
20:37:52.323 | ➤ YN0000: │ /tmp/xfs-76265625 STDOUT Some issues need review, and may require choosing
-- | --
20:37:52.323 | ➤ YN0000: │ /tmp/xfs-76265625 STDOUT a different dependency.
20:37:52.323 | ➤ YN0000: │ /tmp/xfs-76265625 STDOUT
20:37:52.324 | ➤ YN0000: │ /tmp/xfs-76265625 STDOUT Run `npm audit` for details.
20:37:52.869 | ➤ YN0058: │ next-on-pages@https://github.com/cloudflare/next-on-pages.git#commit=5bfbbe15ce14369016992a7cca9f93847851be67: Packing the package failed (exit code 1, logs can be found here: /tmp/xfs-bf62a741/pack.log)
20:37:52.870 | ➤ YN0000: └ Completed in 1m 46s
20:37:52.870 | ➤ YN0000: Failed with errors in 1m 46s
20:37:53.996 | Error: Exit with error code: 1
20:37:53.997 | at ChildProcess.<anonymous> (/snapshot/dist/run-build.js)
20:37:53.997 | at Object.onceWrapper (node:events:652:26)
20:37:53.997 | at ChildProcess.emit (node:events:537:28)
20:37:53.998 | at ChildProcess._handle.onexit (node:internal/child_process:291:12)
20:37:54.007 | Failed: build command exited with code: 1
20:37:55.100 | Failed: error occurred while running build command
```

added like so 
```
    "next-on-pages": "git+https://github.com/cloudflare/next-on-pages.git#5bfbbe15ce14369016992a7cca9f93847851be67",
```

**Comment by vans163** [comment]
Does anyone know how to force next-on-pages to be commit or branch bound?

```


20:58:32.965 | Executing user command: npx @cloudflare/next-on-pages@1
-- | --
20:58:35.048 | npm WARN exec The following package was not found and will be installed: @cloudflare/next-on-pages@1.5.0
```

even with this line in package.json it still pulls wrong version

```
 "next-on-pages": "git+https://github.com/cloudflare/next-on-pages.git#5bfbbe15ce14369016992a7cca9f93847851be67",
```

**Comment by dario-piotrowicz** [comment]
@vans163 is there a reason why you can't use the PR's [prerelase](https://github.com/cloudflare/next-on-pages/pull/418#issuecomment-1667921974)?

Also since the change made it into main it's also been published in the latest [beta](https://www.npmjs.com/package/@cloudflare/next-on-pages/v/0.0.0-705a95b)

**Comment by vans163** [comment]
> @vans163 is there a reason why you can't use the PR's [prerelase](https://github.com/cloudflare/next-on-pages/pull/418#issuecomment-1667921974)?
> 
> Also since the change made it into main it's also been published in the latest [beta](https://www.npmjs.com/package/@cloudflare/next-on-pages/v/0.0.0-705a95b)

I did `npm i @cloudflare/next-on-pages@0.0.0-705a95b`

```
  "dependencies": {
    "@cloudflare/next-on-pages": "^0.0.0-705a95b",
```

I dont get why this keeps happening in the cloudflare pipeline

```
09:54:20.388 | Executing user command: npx @cloudflare/next-on-pages@1
-- | --
09:54:22.876 | npm WARN exec The following package was not found and will be installed: @cloudflare/next-on-pages@1.5.0


```

**Comment by emrecoban** [comment]
@vans163 I've checked my app's build logs, and then seen the same warn. You can ignore it.

Just check the `devDependencies` list:
```bash
01:35:41.009 | devDependencies:
01:35:41.009 | + @cloudflare/next-on-pages 0.0.0-54d6325
01:35:41.010 | + @tailwindcss/forms 0.5.4
01:35:41.010 | + vercel 30.0.0
01:35:41.010 |  
01:35:41.010 | Done in 29.2s
01:35:41.050 | Executing user command: npx @cloudflare/next-on-pages@1
01:35:47.290 | npm WARN exec The following package was not found and will be installed: @cloudflare/next-on-pages@1.5.0
```

Everything is okay in the app now.

**Comment by vans163** [comment]
> @vans163 I've checked my app's build logs, and then seen the same warn. You can ignore it.
> 
> Just check the `devDependencies` list:
> 
> ```shell
> 01:35:41.009 | devDependencies:
> 01:35:41.009 | + @cloudflare/next-on-pages 0.0.0-54d6325
> 01:35:41.010 | + @tailwindcss/forms 0.5.4
> 01:35:41.010 | + vercel 30.0.0
> 01:35:41.010 |  
> 01:35:41.010 | Done in 29.2s
> 01:35:41.050 | Executing user command: npx @cloudflare/next-on-pages@1
> 01:35:47.290 | npm WARN exec The following package was not found and will be installed: @cloudflare/next-on-pages@1.5.0
> ```
> 
> Everything is okay in the app now.
> 
> You can install the `54d6325` in your project with:
> 
> ```shell
> npm install --save-dev https://prerelease-registry.devprod.cloudflare.dev/next-on-pages/runs/5786546906/npm-package-next-on-pages-418
> ```

Yea same error

```


10:11:38.298 | ⚡️ Completed `npx vercel build`.
-- | --
10:11:40.688 |  
10:11:40.689 | ⚡️ ERROR: Failed to produce a Cloudflare Pages build from the project.
10:11:40.689 | ⚡️
10:11:40.689 | ⚡️ 	The following routes were not configured to run with the Edge Runtime:
10:11:40.689 | ⚡️ 	  - /_error
10:11:40.689 | ⚡️
10:11:40.690 | ⚡️ 	Please make sure that all your non-static routes export the following edge runtime route segment config:
10:11:40.690 | ⚡️ 	  export const runtime = 'edge';
10:11:40.690 | ⚡️
10:11:40.690 | ⚡️ 	You can read more about the Edge Runtime on the Next.js documentation:
10:11:40.690 | ⚡️ 	  https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes
10:11:40.690 |  
10:11:40.732 | Failed: Error while executing user command. Exited with error code: 1
10:11:40.743 | Failed: build command exited with code: 1
10:11:44.924 | Failed: error occurred while running build command
```

This is the repo, maybe if you have a chance to look I would really appreciate it. I have no idea whats wrong. It is using the app/ instead of pages/ feature of NEXTJS.  Its a charity project to help out a not-for-profit.

https://github.com/vans163/link

**Comment by emrecoban** [comment]
@vans163 I've tried several ways, but couldn't be successful. I am not sure what causes this error. Although I added different versions of next-on-pages (included beta), it threw the same error. I should say that your packages are not working together. Maybe it is related to that.

```bash
 WARN  Issues with peer dependencies found
.
└─┬ pliny 0.1.1
  └─┬ kbar 0.1.0-beta.41
    └─┬ react-virtual 2.10.4
      └── ✕ unmet peer react@"^16.6.3 || ^17.0.0": found 18.2.0
```

**Comment by vans163** [comment]
> @vans163 I've tried several ways, but couldn't be successful. I am not sure what causes this error. Although I added different versions of next-on-pages (included beta), it threw the same error. I should say that your packages are not working together. Maybe it is related to that.
> 
> ```shell
>  WARN  Issues with peer dependencies found
> .
> └─┬ pliny 0.1.1
>   └─┬ kbar 0.1.0-beta.41
>     └─┬ react-virtual 2.10.4
>       └── ✕ unmet peer react@"^16.6.3 || ^17.0.0": found 18.2.0
> ```

So moving the project out was gonna be a pain, I read carefully your PR #418 and hit a eureka. Deleting `not-found` from the `app/` root made it automagically-gandalf-wizardry stop producing the `_error` artifact after compile. It compiled!

**Comment by vans163** [comment]
Now its giving
```
      "message": [
        "EvalError: Code generation from strings disallowed for this context"
      ],

```
visiting any page on the router lol..

**Comment by wuifdesign** [comment]
adding a `not-found.tsx` still gives me the same error. I'm using `@cloudflare/next-on-pages CLI v.1.5.1` but also tried the newest beta without a difference.

Is there any specific setting i have to set to get this working, as above is mentioned someone got a successful build.

```
⚡️ ERROR: Failed to produce a Cloudflare Pages build from the project.
⚡️ 
⚡️      The following routes were not configured to run with the Edge Runtime:
⚡️        - /_error
⚡️ 
⚡️      Please make sure that all your non-static routes export the following edge runtime route segment config:
⚡️        export const runtime = 'edge';
⚡️ 
⚡️      You can read more about the Edge Runtime on the Next.js documentation:
⚡️        https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes
```

**Comment by wuifdesign** [comment]
@emrecoban i found my problem. i used the `generateMetadata` in the layout. seams like the layout also have to be static if using a not found.

**Comment by wuifdesign** [comment]
i would have a question about this static rendered 404 page: 
I'm using CSP nonce for security reasons and it works fine in the whole app, but not for the 404, as the nonce isn't injected to the code of the 404 pages (as they are static), but the middleware is running and adding the header which results in CSP errors.

I implemented CSP like in this example https://github.com/cloudflare/next-on-pages/discussions/442

**Comment by yongyi520** [comment]
> Now its giving
> 
> ```
>       "message": [
>         "EvalError: Code generation from strings disallowed for this context"
>       ],
> ```
> 
> visiting any page on the router lol..
> 
> EDIT: This is most likely due to the mdx templating engine to turn markdown into html code. Dead end it seems.

Were you able to fix this? I also got the same issue when doing pdx templating engine with contentlayer

**Comment by anonymouscatcher** [comment]
@dario-piotrowicz is this fixed under certain branch/pre-release? I tried with your branch but still happening

**Comment by tao** [comment]
I was able to get the app to build. I just added the `not-found.tsx` file to the app folder to resolve the `/_not-found` bug.

```tsx
// app/not-found.tsx
export default function NotFound() {
  return (
    <div>
       <h2>Not Found</h2>
       <p>Could not find requested resource</p>
      <a href="/">Return Home</a>
    </div>
  )
}
```

Then to overcome the `_error` issue, I couldn't fix it with the app router custom errors but I just added custom errors to the `pages` folder.

```bash
pages
├── 404.js
└── 500.js
```

These files are just very simple for now:

```tsx
// pages/404.js
export default function Custom404() {
  return <h1>404 - Page Not Found</h1>
}
```

```tsx
// pages/500.js
export default function Custom500() {
  return <h1>500 - Server Error</h1>
}
```

Now the app builds with `npx @cloudflare/next-on-pages@1`.

**Comment by cbonaco1** [comment]
the fix @tao mentioned worked for me, in the sense that it got `@cloudflare/next-on-pages` to build. BUT the only problem is that when the user goes to a not found route (ie `/asdf`), the `/pages/404.js` is rendering instead of `/app/not-found.tsx`. I can work around this by just importing the same component used in `/app/not-found.tsx` but this seems hacky especially since I'm trying to use only the app router. 

I guess this proves further that `@cloudflare/next-on-pages` has issues with error and not found pages in the app router when they use the edge runtime, but not for the pages router.

@dario-piotrowicz appreciate anything you can do for this 💪

**Comment by brianfryer** [comment]
> @emrecoban i found my problem. i used the `generateMetadata` in the layout. seams like the layout also have to be static if using a not found.
> 
> But this doesn't show the warning for the dynamic not-found page

This is the case for me as well.

When I added `export const runtime = 'edge';` to `app/layout.jsx`, `@ cloudflare/next-on-pages` would fail to build. When I removed the `runtime` export, builds started working again.

**Comment by azvyae** [comment]
@brianfryer my workaround currently by creating export const runtime edge on the layout, and force the not-found tsx use nodejs runtime, somehow its working

**Comment by brianfryer** [comment]
> @brianfryer my workaround currently by creating export const runtime edge on the layout, and force the not-found tsx use nodejs runtime, somehow its working

🙃. I’m gonna give that a shot today.

**Comment by dario-piotrowicz** [comment]
Basically to recap, I implemented the fix in https://github.com/cloudflare/next-on-pages/pull/418 and it ""fixes"" the not-found issue in the sense that:
 - if you have a static `app/not-found.tsx` it uses it with no problems
 - if you have an `app/not-found.tsx` with server logic then it still breaks, but the eslint-plugin at least warns you about that
 
 (as described [here](https://github.com/dario-piotrowicz/next-on-pages/blob/207de505b9e479eb1f4bff65964b1b0779b7ac1d/.changeset/little-hotels-wave.md))
 
 I just checked and the above works as intended (I just tested it with the [minimal reproduction](https://github.com/dario-piotrowicz/next-on-pages-app-not-found-vercel-31-issue-reproduction) I created, with the latest package versions).

Now the issue that people are experiencing is that it you have a static `app/not-found.tsx` AND an `app/layout.tsx` with server side logic, then the build fails because of the _error issue again 😕 (I checked it [here](https://github.com/dario-piotrowicz/next-on-pages-app-not-found-vercel-31-issue-reproduction/compare/main...with-ssr-layout) and it does break the build unfortunately) , so that's the problem right? that the solution implemented above breaks when you add such an `app/layout.tsx`, am I correct? is there anything else I'm missing or does this sums it up?

___

If the above is correct I am afraid that the only solution will be to say that we can't do anything about that and say that if you've got an `app/not-found.tsx` then also `app/layout.tsx` can't contain server side logic 😓 (and unfortunately I don't know if we'll be able to add an eslint plugin that help such case 😓)

But regardless before throwing in the towel I'll try to explore this and see if anything better can be done 😢

**Comment by gowza** [comment]
It seems like not found and error pages/routes are auto generated on build when you don't define `app/not-found.tsx`.

I don't have `app/not-found.tsx` and I had to remove server side logic from `app/layout.tsx` to make the build work again.

I'm on Next.js version 13.5.4, using only App Router, and building with `npx @cloudflare/next-on-pages@1`

**Comment by microooji** [comment]
> It seems like not found and error pages/routes are auto generated on build when you don't define `app/not-found.tsx`.
> 
> I don't have `app/not-found.tsx` and I had to remove server side logic from `app/layout.tsx` to make the build work again.
> 
> I'm on Next.js version 13.5.4, using only App Router, and building with `npx @cloudflare/next-on-pages@1`

could you write example app/layout.tsx?

**Comment by gowza** [comment]
@microooji Sure. Example `app/layout.tsx`:

```
import './globals.css'
import { Inter } from 'next/font/google'
import { cookies } from 'next/headers';

const inter = Inter({ subsets: ['latin'] })

export const runtime = "edge"; // <--  Specifying edge runtime.

export const metadata = {
  title: 'Create Next App',
  description: 'Generated by create next app',
}

export default function RootLayout({
  children,
}: {
  children: React.ReactNode
}) {
  console.log(cookies().get('some-cookie')); // <-- reading cookies turning route into dynamic rendering.

  return (
    <html lang="en">
      <body className={inter.className}>{children}</body>
    </html>
  )
}
```

Without `app/not-found.tsx`, this is the error I see:
![image](https://github.com/cloudflare/next-on-pages/assets/7310270/5987f583-7884-41a5-8f3b-9dbaec724682)

**Comment by fforres** [comment]
Been having the same issue as you folks mentioned.

What ended up working for me now (PS: I'm using `src/app` folder):

- Have a `src/app/error.tsx` with `"use client"` directive
- Have a `src/app/not-found.tsx` with `"use client"` directive
- Have a `src/pages/404.js` (No directive)
- Have a `src/pages/404.js` (No directive)
- Specify the edge runtime in my `src/app/layout.tsx`.

<img width="893" alt="image" src="https://github.com/cloudflare/next-on-pages/assets/952992/f5f5a6b3-3010-4cef-b951-b5d328cdae7f">

-----------------

Aside from that, the usuals. :) 

If it helps. This is a permalink to the commit that worked in our project 😄 
https://github.com/JSConfCL/turuleca/tree/c9b2060aab57f7d0c0bb5b56ff56f6ffa231c10a


![image](https://github.com/cloudflare/next-on-pages/assets/952992/3510a6e6-3833-4ba8-9901-6469d6bcd62d)

**Comment by fforres** [comment]
Though now we are getting some client-side errors. https://ad924d6e.turuleca.pages.dev/

`Error: NextRouter was not mounted`

`Application error: a client-side exception has occurred (see the browser console for more information).`

**Comment by arthureberledev** [comment]
After trying several things the only solution that worked for me was to remove `export const runtime = "edge"` from both my `not-found.tsx` and `app/layout.tsx`.

**Comment by YoranBrondsema** [comment]
> Been having the same issue as you folks mentioned.
> 
> What ended up working for us. Which fixes the deploy, but still fails a runtime
> 
>     * Have a `src/app/error.tsx` with `"use client"` directive
> 
>     * Have a `src/app/not-found.tsx` with `"use client"` directive
> 
>     * Have a `src/pages/404.js` (No directive)
> 
>     * Have a `src/pages/404.js` (No directive)
> 
>     * Specify the edge runtime in my `src/app/layout.tsx`.
> 
> 
> (PS: I'm using `src/app` folder):
> <img alt="image" width="893" src="https://user-images.githubusercontent.com/952992/273437750-f5f5a6b3-3010-4cef-b951-b5d328cdae7f.png">
> 
> Aside from that, the usuals. :)
> 
> If it helps. This is a permalink to the commit that worked in our project 😄 https://github.com/JSConfCL/turuleca/tree/c9b2060aab57f7d0c0bb5b56ff56f6ffa231c10a
> 
> ![image](https://user-images.githubusercontent.com/952992/273437932-3510a6e6-3833-4ba8-9901-6469d6bcd62d.png)
> 
> Though as mentioned, now we are getting some client-side errors. [ad924d6e.turuleca.pages.dev](https://ad924d6e.turuleca.pages.dev/)
> 
> Console
> 
> ```
> Error: NextRouter was not mounted
> ```
> 
> <img alt="" width="300px" src="https://user-images.githubusercontent.com/952992/273438130-f4a7de05-f0a0-46e6-b6ab-31952cba0083.png">

FYI this worked for us. Thanks!

**Comment by fforres** [comment]
Realized that my issue is most likely due to middleware.ts and the clerk provider. WIll ideally work on it and see if it solved. Glad i helped @YoranBrondsema

**Comment by vader1359** [comment]
Guys I have this error with Page Router, too. not sure if it's the same error.
```

⚡️ ERROR: Failed to produce a Cloudflare Pages build from the project.
--
23:52:13.699 | ⚡️
23:52:13.699 | ⚡️ 	The following routes were not configured to run with the Edge Runtime:
23:52:13.699 | ⚡️ 	  - /_next/data/T2unbIx1RCCAyFtaJjxro/brands/[...id].json
23:52:13.699 | ⚡️ 	  - /_next/data/T2unbIx1RCCAyFtaJjxro/designers/[...id].json
23:52:13.699 | ⚡️ 	  - /_next/data/T2unbIx1RCCAyFtaJjxro/news/[...id].json
23:52:13.699 | ⚡️ 	  - /_next/data/T2unbIx1RCCAyFtaJjxro/products/[...id].json
23:52:13.699 | ⚡️ 	  - /_next/data/T2unbIx1RCCAyFtaJjxro/products/category/[...category].json
23:52:13.699 | ⚡️ 	  - /_next/data/T2unbIx1RCCAyFtaJjxro/products/sub-category/[...sub-category].json
23:52:13.700 | ⚡️ 	  - /brands/[...id]
23:52:13.700 | ⚡️ 	  - /designers/[...id]
23:52:13.700 | ⚡️ 	  - /products/[...id]
23:52:13.700 | ⚡️ 	  - /products/category/[...category]
23:52:13.700 | ⚡️ 	  - /products/sub-category/[...sub-category]
23:52:13.700 | ⚡️
23:52:13.701 | ⚡️ 	Please make sure that all your non-static routes export the following edge runtime route segment config:
23:52:13.701 | ⚡️ 	  export const runtime = 'edge';

```

All these paths have getStaticPaths function.
- I tried to add `export const runtime = 'edge'` to these pages, it will complain `Error: Page /news/[...id] provided runtime 'edge', the edge runtime for rendering is currently experimental. Use runtime 'experimental-edge' instead.`
- IF I swith to experimental-edge, it will complain `> webpack config.resolve.alias was incorrectly overridden. https://nextjs.org/docs/messages/invalid-resolve-alias`
- Tried to add the `export const runtime = 'edge'` to layout, it build OK but failed on deployment.

**Comment by dario-piotrowicz** [comment]
@vader1359 I'm pretty sure you're experiencing something different, would it be possible for you to create a minimal reproduction of it to share and open a new issue for it?

**Comment by reiranhr** [comment]
What ended up working for me was:
- Removing `export const runtime = 'edge'` from the `app/layout.tsx` file
- Adding `export const runtime = 'edge'` to each page/routes and to `app/not-found.tsx`

**Comment by kev-ac** [comment]
I get the error message about the missing edge runtime on the _not-found and _error route (both files do not exist in this project) when I try to work with `next/headers` in the layout.tsx file. 
Once I remove this everything is fine.

**Comment by anonymouscatcher** [comment]
@dario-piotrowicz Any updates? do you think this is going to be fixed soon?

**Comment by dario-piotrowicz** [comment]
@anonymouscatcher this issue seems caused by how Vercel builds functions and it is a bit of our control to fix, one possible solution I thought of was allowing users to ask next-on-pages to ignore invalid routes as sometimes those are actually ok to ignore (like in the layout/not-found case mentioned in this thread and reproduced in this [application](https://github.com/dario-piotrowicz/next-on-pages-ignore-error-layout-example)).

I implemented that here: https://github.com/cloudflare/next-on-pages/pull/497, the PR hasn't made much progress since I was trying to get the team's feedback on this since it's really not a super elegant solution.

if you could, could you just the [prerelease](https://github.com/cloudflare/next-on-pages/pull/497#issuecomment-1752701164) a try and let me know what you think? I would surely appreciate external feedback on this.

If the solution above works reasonably well for people I think I can push for it with the team and merge it soon, so again I'd be very thankful if people could give it a go and let me know if it works for them :slightly_smiling_face:

**Comment by maghfoor-dev** [comment]
Hey @dario-piotrowicz I tried installing your prelease and making it build, but unfortunately still get the
```
The following routes were not configured to run with the Edge Runtime:
⚡️        - /_error
```
Error :(

**Comment by dario-piotrowicz** [comment]
@maghfoor-dev but you also should get there a message saying that you can ignore that route right?

Did you try then to add `-g=_error`?

**Comment by maghfoor-dev** [comment]
@dario-piotrowicz 
I didn't try to do that sorry, do you mean adding that in the build command?
I fixed this by trying to find out what actually was causing the issue. For me it was the `layout.tsx` page being an async component because I wanted to pass the `auth` `SessionProvider` to my app. I removed that and my project built. Weird behaviour but yeah

**Comment by minddust** [comment]
My personal take: 

if `_error` gets falsely interpreted/generated as route, the solution should not the to ignore that entry, but to not generate it in the first place.

Wouldn't that be a bug/responsibility within vercel? Didn't see any issues there mentioning this. (my 4th day of next - don't know)

**Comment by jhd4tmr** [comment]
The above solution did not solve my problem. There is still an _error problem. I have temporarily given up deploying to cloudflare.

**Comment by james-elicx** [comment]
Hi all, this popped into my head just earlier, but you can probably do something like the following for your build command as a temporary workaround. The real fix would be Vercel resolving this.

```sh
npx vercel build; rm -rf .vercel/output/functions/_not-found*; rm -rf .vercel/output/functions/_error*; npx @cloudflare/next-on-pages --skip-build
```

**Comment by always-maap** [comment]
removing app/not-found.tsx fixed it temporarily 

https://github.com/always-maap/NIT-Academy

**Comment by dario-piotrowicz** [comment]
Hi all, I've opened #585 to fix this issue, the changes make it so that invalid `_error` functions generated when you have a layout that uses the edge runtime and also a static not-found route get ignored.

You still however cannot have a not-found route using the edge runtime.

Please give the [prerelease](https://github.com/cloudflare/next-on-pages/pull/585#issuecomment-1849644282) a try and let me know if it works for you 🙏

**Comment by dario-piotrowicz** [comment]
PR https://github.com/cloudflare/next-on-pages/pull/585 has closed this issue, although the situation is still not perfect I would keep this issue closed so not to keep the conversation going, making it too diluted/unfocused

What things should work like right now in applications using the App router (in the latest `beta` release and from the next stable release):
 - `not-found` routes need to be static and not have server side logic (as per https://github.com/cloudflare/next-on-pages/pull/418), if they do the build process is expected to fail
 
 - assuming that the app has a __static__ `not-found` route, the app can have a root layout that does have server side (edge) logic (as per https://github.com/cloudflare/next-on-pages/pull/585)

If you see that any of the two points above doesn't work as expected please comment here and we can reopen this issue, if you spot anything else please let's open a new issue focused on that specific problem 🙏 

Sorry for the trouble here, this one is a bit of a painful issue, hopefully the latest changes help the situation 😅

**Comment by nikitaeverywhere** [comment]
@dario-piotrowicz after adding middleware.ts to the bootstrapped default nextjs project (`npx create-next-app@latest`) as per [their docs](https://nextjs.org/docs/pages/building-your-application/configuring/content-security-policy) (literally copy-paste), I've got the following error:

<img width="991" alt="image" src="https://github.com/cloudflare/next-on-pages/assets/4989256/259f81eb-b86c-4898-ba79-2dde06d5f2a3">

And after adding missing pages and marking all as `export const runtime = "edge";`,

![image](https://github.com/cloudflare/next-on-pages/assets/4989256/004c2deb-e9b5-4ba2-8f88-c92e93cf5ee8)

I still get this:

<img width="931" alt="image" src="https://github.com/cloudflare/next-on-pages/assets/4989256/4173b1cb-a294-447d-b25d-583c59627539">

<img width="221" alt="image" src="https://github.com/cloudflare/next-on-pages/assets/4989256/147e32cf-6f85-48de-9759-ec119701d10c">

(latest available)

Does it mean no way to run it on Cloudflare?

**Comment by dario-piotrowicz** [comment]
@xanderim thanks for confirming that the latest beta release does work 🙂 

I've just cut a standard release so using the [latest stable release](https://www.npmjs.com/package/@cloudflare/next-on-pages/v/1.8.3) should work fine now 🤞 

@nikitaeverywhere please give it a try and let me know if it works for you 🙂

**Comment by nikitaeverywhere** [comment]
@xanderim @dario-piotrowicz thanks for the feedback. I changed the build command in CloudFlare page settings, and both "latest" `npx @cloudflare/next-on-pages@1` and `npx @cloudflare/next-on-pages@0.0.0-2b5c8f2` work as of now.

nextjs 14 is however totally incapable of handling CSP but that's unrelated :D 

Thanks.

**Comment by dario-piotrowicz** [comment]
@nikitaeverywhere I'm glad things work for you now 😄 🚀 

Regarding Next.js 14 and CSP, sorry I am not aware of the incompatibility you're mentioning, is that something totally on Next.js? or is it something caused by our adapter? (if it's the latter please open a [new gh issue](https://github.com/cloudflare/next-on-pages/issues/new/choose) for that, so that we can look into it 🙏)

**Comment by nikitaeverywhere** [comment]
> Regarding Next.js 14 and CSP

@dario-piotrowicz it's on the NextJS side completely. I personally think NextJS is just not designed to support CSP. You can find a lot of [lengthy threads](https://github.com/vercel/next.js/discussions/54907) with other people suffering.

So, as for Cloudflare, no worries at all :D

**Comment by dario-piotrowicz** [comment]
I see @nikitaeverywhere I'm glad that it's not something we're causing 😄 (but of course I'm sad that Next.js seem to present this many issues around CSP 🥲)

Anyways thanks a ton for sharing the Next.js issue, I gave it a quick scan through and it does feel like something something that could definitely come in handy to be aware of and keep in mind! so again thanks a bunch! ❤️

**Comment by switz** [comment]
> I still have this issue locally, I have only one not-found.js in top level

seeing this same issue myself, has anyone found a solution? (generally wouldn't bump but its 2.5 months old)

**Comment by cameron-eth** [comment]
Hey all, I've built on top of morphic but everytime the search results post, the app re routes to the home page ?

**Comment by AlecKriebel** [comment]
For others experiencing this issue, upgrading from `"vercel": "^31.1.0"` to `"vercel": "^32.4.1"` (latest) in the Next.js project I was trying to build for pages fixed this for me.

**Comment by DeJayDev** [comment]
Upgrading the `vercel` dependency did not fix the error in my project, however I've seen successful builds by entirely removing the _error directory before running Next on Pages.

`
npm run build; rm -rf .vercel/output/functions/_error*; npx @cloudflare/next-on-pages --skip-build
`

This command was written by @james-elicx in the Cloudflare Developers Discord.

**Comment by dario-piotrowicz** [comment]
Sorry for the late reply (I don't really notice much comments on closed issues 😓)

@anonymouscatcher and @DeJayDev if you experience the `_error` problem, please feel free to open a new issue for it with a minimal reproduction 🙏 (emphasis on the minimal reproduction, since, without one I feel like it's going to be very difficult to get to the bottom of things 😓).

PS: @DeJayDev, please note that the command you shared (from @james-elicx) is (at least in my opinion) not really safe as that completely removes the problematic file without any guarantee that the application will run correctly in production (if you want you can indeed use it, but in that case you'd probably also want to very very thoroughly test that the built application works as intended before deploying it (with `wrangler pages dev`))

**Comment by charlie0077** [comment]
Running into the same issue in 2025.
ERROR: Failed to produce a Cloudflare Pages build from the project.
⚡️
⚡️ 	The following routes were not configured to run with the Edge Runtime:
⚡️ 	  - /_error
⚡️
⚡️ 	Please make sure that all your non-static routes export the following edge runtime route segment config:
⚡️ 	  export const runtime = 'edge';
⚡️
⚡️ 	You can read more about the Edge Runtime on the Next.js documentation:
⚡️ 	  https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes

This issue is closed, but what is the solution?

**Comment by venkateshwaraChinnasamy** [comment]
is there no solution for this? i'm also facing the same issue
